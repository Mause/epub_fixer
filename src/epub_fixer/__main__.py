import re
from functools import cache
from pathlib import Path

import rich_click as click
from ebooklib.epub import read_epub, write_epub
from epubcheck import EpubCheck
from rich import print
from rich.prompt import Prompt


@click.command
@click.argument("filename")
def main(filename: str):
    """
    filename\tthe file to fix
    """
    title = cache(
        lambda: Prompt.ask(
            "Enter a title for this book",
            default=Path(filename).stem,
            show_default=True,
        )
    )

    print("checking for issues")
    result = EpubCheck(filename, autorun=False)
    result.run()

    if not result.messages:
        print("No issues found")
        exit(0)

    book = read_epub(filename, {"ignore_ncx": False})

    for message in result.messages:
        print(message)
        msg = message.message
        if (
            msg
            == 'The "direction" property must not be included in an EPUB Style Sheet.'
        ):
            (item,) = book.get_items_of_media_type("text/css")

            item.content = re.sub(
                r"(direction: [^;]+;)",
                lambda *args: "",
                item.content.decode("utf-8"),
            ).encode("utf-8")
        elif msg.endswith('The "head" element should have a "title" child element.'):
            name, row, col = message.location.split(":")
            name = name.split("/", 2)[-1]
            item = next(i for i in book.items if i.file_name == name)
            if name.endswith("toc.xhtml"):
                item.title = title()
            else:
                item.title = Prompt.ask(
                    "Enter a title for this item",
                    show_default=True,
                )
        else:
            raise Exception(f"Unknown issue: {msg}")

    if book.title == "Unknown Title":
        book.get_metadata("DC", "title")[0] = (title(), {})

    authors = book.get_metadata("DC", "creator")
    if authors[0][0] == "Unknown Author":
        authors[0] = (Prompt.ask("Enter an author for this book"), {})

    fixed = filename + ".fixed.epub"
    write_epub(fixed, book)
    print(f"Fixed book written to {fixed}")


if __name__ == "__main__":
    main()
