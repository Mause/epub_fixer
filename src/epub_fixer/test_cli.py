from os.path import basename
from pathlib import Path
from urllib.request import urlretrieve

from click.testing import CliRunner
from pytest import fixture

from .__main__ import epub_fixer


@fixture
def happy_epub(tmp_path):
    url = "https://github.com/IDPF/epub3-samples/releases/download/20230704/accessible_epub_3.epub"
    return get_epub(tmp_path, url)


def get_epub(tmp_path: Path, url: str) -> Path:
    out = tmp_path / basename(url)
    urlretrieve(url, out)
    return out


def test_all_good(happy_epub):
    runner = CliRunner()

    result = runner.invoke(epub_fixer, [str(happy_epub)])
    assert result.exit_code == 0
    assert result.output == "No issues found\n"


@fixture
def unhappy_epub(tmp_path):
    return get_epub(
        tmp_path,
        "https://github.com/robotmedia/iBooks-ePub-Samples/raw/refs/heads/master/JQuery-Hello-World.epub",
    )


def test_issues(unhappy_epub, snapshot):
    runner = CliRunner()

    result = runner.invoke(epub_fixer, [str(unhappy_epub)])
    assert result.exit_code == 1
    assert result.output == snapshot
