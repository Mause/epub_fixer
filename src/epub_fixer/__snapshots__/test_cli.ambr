# serializer version: 1
# name: test_issues
  '''
  Message(
      id='NCX-001',
      level='ERROR',
      location='JQuery-Hello-World.epub/OEBPS/toc.ncx',
      message='NCX identifier ("e37ff260-d35c-11e0-9572-0800200c9a66") does not 
  match OPF identifier ("urn:uuid:e37ff260-d35c-11e0-9572-0800200c9a66").',
      suggestion=None
  )
  
  '''
# ---
